### ISSUE TITLE
Source etd may crash in remote-remote transfer if etc is ^C aborted quickly

### ISSUE THREAD

**Issue by varenius** [jive-vlbi/etransfer#8]
I tested remote-remote transfers using the latest socket-options branch. Setup: I run etc on "gyller", transferring from machine "skirner_x" to "TARGET". On the machine "skirner_x" I start etd serveras :
```
oper@skirner /usr/local/src/etransfer.git/Linux-x86_64-native-opt $ ./etd --command tcp://:4004 --data udt://:8008
```
Confirm it is running:
```
oper@skirner /usr/local/src/etransfer.git/Linux-x86_64-native-opt $ ps -u oper | grep etd
31418 ?        00:00:00 etd
```

On other machine (gyller), run etc command to transfer from skirner_x to TARGET:
```
oper@gyller:~/eskil/etransfer.git/Linux-x86_64-native-opt$ ./etc skirner_x:/mnt/vbsmnt/ev0244_ow_244-1000* TARGET#2620:/gpfs/cdata/incoming/onsala-test/ -v --resume
2021-05-20 08:58:46.88: PUSH Resume /mnt/vbsmnt/ev0244_ow_244-1000_0 -> /gpfs/cdata/incoming/onsala-test/ev0244_ow_244-1000_0
2021-05-20 08:58:46.92: Destination is complete or is larger than source file
2021-05-20 08:58:46.95: PUSH Resume /mnt/vbsmnt/ev0244_ow_244-1000_1 -> /gpfs/cdata/incoming/onsala-test/ev0244_ow_244-1000_1
2021-05-20 08:58:46.98: Destination is complete or is larger than source file
2021-05-20 08:58:47.01: PUSH Resume /mnt/vbsmnt/ev0244_ow_244-1000_2 -> /gpfs/cdata/incoming/onsala-test/ev0244_ow_244-1000_2
```
so this appears to work. I wait 5 seconds, then abort with ^C. Client stops. On "skirner_x" (the data source), etd is still running. All good.

HOWEVER, if I do the same exercise again, but instead of waiting 5 seconds, I ^C the client (on gyller) almost immediately (after 1 second?), then the etd server (on skirner_x) will crash with this message:
```
oper@skirner /usr/local/src/etransfer.git/Linux-x86_64-native-opt $ terminate called after throwing an instance of 'std::runtime_error'
  what():  udtsend(101423127, .., n=33554432 ..)/Operation not supported: Invalid socket ID. (5004)

```
I am not sure if this is a "short time since start" or "unlucky timing relative to some internal action sequence" problem, but I manageded to reproduce it almost every time when aborting the client with ^C very quickly after starting the command.

**Comment by varenius** [comment]
Running my "etd" command with "-f -m 5" I get this output on screen which may provide a clue:
```
oper@skirner /usr/local/src/etransfer.git/Linux-x86_64-native-opt $ ./etd --command tcp://:4004 --data udt://:8008 -f -m 5
2021-05-20 09:11:48.53: [void signal_thread(const signallist_type&, std::promise<int>&)] sigwaiterthread: enter wait phase
2021-05-20 09:11:48.53: [etdc::etdc_fdptr string2socket_type_m::operator()(const string&) const] etd: server is-at <udt/0.0.0.0:8008/mss=1500,max-bw=-1>
2021-05-20 09:11:48.53: [<lambda(etdc::etdc_fdptr, const string&)>] server/UDT rcvbuf = 33554240 sndbuf = 335543872 linger=0:0 mss=1500 max_bw=Inf
2021-05-20 09:11:48.53: [etdc::etdc_fdptr string2socket_type_m::operator()(const string&) const] etd: server is-at <tcp/0.0.0.0:4004>
2021-05-20 09:11:48.53: [<lambda(etdc::etdc_fdptr, const string&)>] server/TCP rcvbuf = 67108864 sndbuf = 67108864
2021-05-20 09:11:57.80: [void command_server_thread(etdc::etdc_fdptr, etdc::etd_state&) [with int <anonymous> = 10; etdc::etdc_fdptr = std::shared_ptr<etdc::etdc_fd>]] Incoming COMMAND from <tcp/10.100.0.15:42226> [local <tcp/10.100.0.11:4004>]
2021-05-20 09:11:57.80: [<lambda(etdc::etdc_fdptr, const string&)>] client/TCP rcvbuf = 67108864 sndbuf = 67108864
2021-05-20 09:11:57.80: [etdc::ETDServer::ETDServer(etdc::etd_state&)] ETDServer starting, my uuid=tS3pT0TPMb4fHCGL
2021-05-20 09:11:57.80: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-05-20 09:11:57.80: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=17 => nTotal=17
2021-05-20 09:11:57.80: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'protocol-version'
2021-05-20 09:11:57.80: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK 1'
2021-05-20 09:11:57.80: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-05-20 09:11:57.86: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=37 => nTotal=37
2021-05-20 09:11:57.86: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'list /mnt/vbsmnt/ev0244_ow_244-1000*'
2021-05-20 09:11:57.89: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK /mnt/vbsmnt/ev0244_ow_244-1000_0'
2021-05-20 09:11:57.89: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK /mnt/vbsmnt/ev0244_ow_244-1000_1'
2021-05-20 09:11:57.89: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK /mnt/vbsmnt/ev0244_ow_244-1000_2'
2021-05-20 09:11:57.89: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK /mnt/vbsmnt/ev0244_ow_244-1000_3'
2021-05-20 09:11:57.89: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK /mnt/vbsmnt/ev0244_ow_244-1000_4'
2021-05-20 09:11:57.89: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK /mnt/vbsmnt/ev0244_ow_244-1000_5'
2021-05-20 09:11:57.89: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK /mnt/vbsmnt/ev0244_ow_244-1000_6'
2021-05-20 09:11:57.89: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK /mnt/vbsmnt/ev0244_ow_244-1000_7'
2021-05-20 09:11:57.89: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK'
2021-05-20 09:11:57.89: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-05-20 09:11:57.95: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=54 => nTotal=54
2021-05-20 09:11:57.95: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'read-file 2539694560 /mnt/vbsmnt/ev0244_ow_244-1000_0'
2021-05-20 09:11:57.95: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {}; std::__cxx11::string = std::__cxx11::basic_string<char>]] open_file/npath='/mnt/vbsmnt/ev0244_ow_244-1000_0'
2021-05-20 09:11:57.96: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'Remain:0'
2021-05-20 09:11:57.96: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'UUID:tS3pT0TPMb4fHCGL'
2021-05-20 09:11:57.96: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK'
2021-05-20 09:11:57.96: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-05-20 09:11:57.99: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=29 => nTotal=29
2021-05-20 09:11:57.99: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'remove-uuid tS3pT0TPMb4fHCGL'
2021-05-20 09:11:57.99: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: removeUUID(tS3pT0TPMb4fHCGL yields 1
2021-05-20 09:11:57.99: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK'
2021-05-20 09:11:57.99: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-05-20 09:11:58.02: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=54 => nTotal=54
2021-05-20 09:11:58.02: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'read-file 2539316256 /mnt/vbsmnt/ev0244_ow_244-1000_1'
2021-05-20 09:11:58.02: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {}; std::__cxx11::string = std::__cxx11::basic_string<char>]] open_file/npath='/mnt/vbsmnt/ev0244_ow_244-1000_1'
2021-05-20 09:11:58.02: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'Remain:0'
2021-05-20 09:11:58.02: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'UUID:tS3pT0TPMb4fHCGL'
2021-05-20 09:11:58.02: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK'
2021-05-20 09:11:58.02: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-05-20 09:11:58.05: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=29 => nTotal=29
2021-05-20 09:11:58.05: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'remove-uuid tS3pT0TPMb4fHCGL'
2021-05-20 09:11:58.05: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: removeUUID(tS3pT0TPMb4fHCGL yields 1
2021-05-20 09:11:58.05: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK'
2021-05-20 09:11:58.05: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-05-20 09:11:58.08: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=54 => nTotal=54
2021-05-20 09:11:58.08: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'read-file 2084099408 /mnt/vbsmnt/ev0244_ow_244-1000_2'
2021-05-20 09:11:58.08: [int etdc::detail::open_file(const string&, int, Args&& ...) [with Args = {}; std::__cxx11::string = std::__cxx11::basic_string<char>]] open_file/npath='/mnt/vbsmnt/ev0244_ow_244-1000_2'
2021-05-20 09:11:58.10: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'Remain:455570480'
2021-05-20 09:11:58.10: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'UUID:tS3pT0TPMb4fHCGL'
2021-05-20 09:11:58.10: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: sending reply 'OK'
2021-05-20 09:11:58.10: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-05-20 09:11:58.10: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=100 => nTotal=100
2021-05-20 09:11:58.10: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'send-file tS3pT0TPMb4fHCGL OuOCOezxWHgzSWlrXL 455570480 <udt/193.170.79.54:2630/mss=1500,max-bw=-1>'
2021-05-20 09:11:58.10: [etdc::sockname_type etdc::decode_data_addr(const string&)] decode_data_addr: 1='udt' 2='193.170.79.54' 9='2630' 11='mss=1500,max-bw=-1'
2021-05-20 09:11:58.10: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-05-20 09:11:58.10: [etdc::ETDServerWrapper::handle()::<lambda()>] ETDServerWrapper: thread 140482692749056/executing sendFile()
2021-05-20 09:11:58.10: [virtual etdc::xfer_result etdc::ETDServer::sendFile(const etdc::uuid_type&, const etdc::uuid_type&, off_t, const dataaddrlist_type&)] ETDServer::sendFile/use MSS=1500 [ours=0, 193.170.79.54=1500]
2021-05-20 09:11:58.10: [virtual etdc::xfer_result etdc::ETDServer::sendFile(const etdc::uuid_type&, const etdc::uuid_type&, off_t, const dataaddrlist_type&)] ETDServer::sendFile/use MaxBW=-1 [ours=0, 193.170.79.54=-1]
2021-05-20 09:11:58.10: [etdc::etdc_fdptr mk_client(const T&, const etdc::detail::client_settings&) [with T = etdc::protocol_type; etdc::etdc_fdptr = std::shared_ptr<etdc::etdc_fd>]] mk_client/attempt #1/3 trying to connect to udt:193.170.79.54:2630
2021-05-20 09:11:58.16: [virtual etdc::xfer_result etdc::ETDServer::sendFile(const etdc::uuid_type&, const etdc::uuid_type&, off_t, const dataaddrlist_type&)] sendFile/connected to <udt/193.170.79.54:2630/mss=1500,max-bw=-1>
2021-05-20 09:11:58.96: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=24 => nTotal=24
2021-05-20 09:11:58.96: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'cancel tS3pT0TPMb4fHCGL'
2021-05-20 09:11:58.96: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper: canelling UUID tS3pT0TPMb4fHCGL
2021-05-20 09:11:58.96: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / start loop, curPos=0
2021-05-20 09:11:59.07: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle() / read n=29 => nTotal=29
2021-05-20 09:11:59.07: [void etdc::ETDServerWrapper::handle()] ETDServerWrapper::handle()/got line: 'remove-uuid tS3pT0TPMb4fHCGL'
terminate called after throwing an instance of 'std::runtime_error'
  what():  udtsend(1034315234, .., n=33554432 ..)/Operation not supported: Invalid socket ID. (5004)
Aborted
```

**Comment by haavee** [comment]
Hmmm I could reproduce it exactly _once_ (the first time). After that I couldn't make the source `etd` crash no matter what. I also encountered an instance where `^C` didn't work at all.

**Comment by haavee** [comment]
This commit https://github.com/jive-vlbi/etransfer/commit/afd069a5e7e07a7224eb551c0429e9625752c2db seems to fix issue #8

**Comment by varenius** [comment]
Testing with the new commit, I cannot make it fail again. So it seems this issue is indeed fixed.
